### DISCUSSION TITLE
Make snippet available only if `TM_SELECTED_TEXT` has a value

### DISCUSSION THREAD

**Discussion by Samasaur1** [L3MON4D3/LuaSnip#647] [Q&A]
Let's say I have a Swift snippet called `printv`, with the end result looking like this:
```swift
print("myvar=\(myvar)")
```

I want to be able to use it three separate ways:

1. Standalone
```lua
    s({
        trig="printv",
        name="printv-standalone",
        dscr="Print [type here]=<value>",
    }, {
        t("print(\""),
        rep(1),
        t("=\\("),
        i(1),
        t(")\")")
    }),
```
You type `printv` on an empty line and it expands to let you type in the name of the variable, which it then duplicates.
2. With selection
```lua
    s({
        trig="printv",
        name="printv-selection",
        dscr="Print 'selected variable'=<value>",
    }, {
        t('print("'),
        f(function(_, snip)
            -- TM_SELECTED_TEXT is a table to account for multiline-selections.
            --          -- In this case only the first line is inserted.
            return snip.env.TM_SELECTED_TEXT[1] or {}
        end, {}),
        t('=\\('),
        f(function(_, snip)
            -- TM_SELECTED_TEXT is a table to account for multiline-selections.
            --          -- In this case only the first line is inserted.
            return snip.env.TM_SELECTED_TEXT[1] or {}
        end, {}),
        t(')")')
    }),
```
Select an expression (e.g., `viw`), hit Tab (in my case, since
```lua
luasnip.config.set_config({
    store_selection_keys="<Tab>",
})
```
), then type `printv` and it will fill in the selection.
3. Postfix
I haven't written it yet, but I don't think it will be too complicated.

***

My question is about trying to get only one of these to show up at a time. If you type `printv` on a blank line, both `printv-standalone` and `printv-selection` show up. I'd like to have them set up so that, if `TM_SELECTED_TEXT` is empty, I only see `printv-standalone`, and if `TM_SELECTED_TEXT` has a value, I only see `printv-selection`. I'm not sure exactly how a postfix version will fit into the mix, because I haven't written that yet either, but I'd like it to also be disambiguated.

Is this possible?

**Comment by L3MON4D3** [comment]
Oh, nice, that's definitely possible. I'll list some ingredients :D
* [`condition`/`show_condition`](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#snippets) can hide a snippet (first for regular expansion, second for completion engines)
* You can basically just write a show_condition and pass it to both options (they're compatible with both), oviously negated for one snippet (recommending [`condition_objects`](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#condition-objects) for doing that in a readable fashion)
* There's currently no api for getting the selection that would be consumed by the next snippet, but you can just rely on the [implementation](https://github.com/L3MON4D3/LuaSnip/blob/22b7c117de19a8de815823e5c32cbbd551d0cd10/lua/luasnip/util/util.lua#L265-L285), it's unlikely to change (Basically, the selection is stored in a buffer-local variable, you can just check if it exists via `pcall` like in the linked function) (don't call `get_selection` itself though, it will consume the selection (is inappropriately named, `s/get/consume/` :sweat_smile:))

Alternative approach: wrap the entire snippet in a dynamicNode which checks if there was selected text, and returns one or the other (could also be used for postfix, by checking its variable as well)

Oh or!: use `TM_SELECTED_TEXT` as the preset text for the insertNodes in the first snippet, then you should also have all in one, I think, and could do it with just `parse`: `"visual or empty: ${1:$TM_SELECTED_TEXT}"`

**Reply by Samasaur1** [reply]
Hmm. It makes sense what I'm _trying_ to do here, but for some reason it never seems to be getting the selection right
```lua
    s({
        trig="pv",
        name="printv dyn",
        dscr="pv",
    }, d(1, function(args)
        local ok, val = pcall(vim.api.nvim_buf_get_var, 0, "LUASNIP_SELECT_RAW");
        if ok then
            return sn(nil, {
                t('print("'),
                f(function(_, snip)
                    return snip.env.TM_SELECTED_TEXT[1] or {}
                end, {}),
                t('=\\('),
                f(function(_, snip)
                    return snip.env.TM_SELECTED_TEXT[1] or {}
                end, {}),
                t(')")')
            });
        else
            return sn(nil, {
                t("print(\""),
                rep(1),
                t("=\\("),
                i(1),
                t(")\")")
            });
        end
    end, nil)
    ),
```

Even when I select something first, and then type `pv` and expand the snippet, it still goes with the `else` clause, and I don't know why.

**Reply by Samasaur1** [reply]
I think I can get the standalone and selection versions to only appear in non-postfix contents by changing it to a `regTrig` (I'm not sure which postfix variable you're talking about to check)

**Reply by L3MON4D3** [reply]
> Even when I select something first, and then type pv and expand the snippet, it still goes with the else clause, and I don't know why.

Mhmm, that's weird, I can't spot anything wrong.. try verifying that the selection is actually populated

**Reply by L3MON4D3** [reply]
> (I'm not sure which postfix variable you're talking about to check)

I meant `POSTFIX_MATCH`, but I was thinking completely along the wrong lines :sweat_smile:

**Reply by Samasaur1** [reply]
> > Even when I select something first, and then type pv and expand the snippet, it still goes with the else clause, and I don't know why.
> 
> Mhmm, that's weird, I can't spot anything wrong.. try verifying that the selection is actually populated

I'm pretty sure the selection is actually populated — my other snippets that use selection still work. Here's a GIF of me using this new `dynamicNode` based snippet, which doesn't work for selections, compared to the working selection snippet above.
![GIF demonstrating usage](https://user-images.githubusercontent.com/30577766/198960969-e993df79-bd39-4e16-ae32-1dbf0ffb28f8.gif)

**Reply by L3MON4D3** [reply]
weird.. Try monitoring the returned values via print (in the dynamicNode)

**Comment by Samasaur1** [comment]
For completeness, here's the postfix version:
```lua
    require("luasnip.extras.postfix").postfix(".printv", {
        t("print(\""),
        l(l.POSTFIX_MATCH),
        t("=\\("),
        l(l.POSTFIX_MATCH),
        t(")\")"),
        i(0)
    }),
```
